_posts/2020-08-03-the-missing-combine-input-type.md

```diff
@@ -0,0 +1,368 @@
+class AuthViewModel {
+    // MARK: - Outputs
+    let signInResult: AnyPublisher<Result<Token, Error>, Never>
+
+    // MARK: - Lifecycle
+    init() {
+        //...
+    }
+}
+```
+
+Here we have `signInResult` using the `Result` type as its output. When authentication is successful it will emit a `.success(Token)` value, 
+when it fails it will output a `.failure(Error)` value.
+
+So how do we go about _triggering_ the work and eventually get a value from this output? We need an input!
+
+## Inputs... ?
+
+But didn't we already decide Combine doesn't have any input types? It _does_ have ways to model inputs... they just come with some baggage.
+
+The types Combine gives us to send values are called `Subject`s. _But_, they are _both_ input _and_ output (`Publisher`). 
+This means that while we can use them to trigger our authentication attempt we need to be careful about how we use them if we care about encapsulation.
+
+Let's introduce a `Subject` and get our view model actually doing something:
+
+```swift
+class AuthViewModel {
+    // MARK: - Outputs
+    let signInResult: AnyPublisher<Result<Token, Error>, Never>
+
+    // MARK: - Inputs
+    let signIn = PassthroughSubject<(username: String, password: String), Never>()
+
+    // MARK: - Lifecycle
+    init() {
+        signInResult = signIn
+            .flatMap { input in
+                return API
+                    .authenticate(input.username, input.password)
+                    .materialize()
+            }
+            .eraseToAnyPublisher()
+    }
+}
+```
+
+A `PassthroughSubject` does exactly what it  says on the label.  It _passes_ values _through_ that it receives via it's `send` function fulfilling its function as an _input_. It then uses those values as the source for its function as a `Publisher`, or _output_.
+
+Now we have defined an input we can use it to attempt to authenticate and, finally, emit the result through our output. Let's take a look at what some code using our view model might look like:
+
+```swift
+let viewModel = AuthViewModel()
+
+viewModel
+    .signInResult
+    .sink { result in
+        print(result) // output: success(Token(value: "foobar"))
+    }
+    .store(in: &cancellables)
+
+viewModel.signIn.send((username: "iankeen", password: "super_secret_password")) // input
+```
+
+Not bad..., however remember `Subject`s are both input _and_ output. This means we have also left the door open for code like this:
+
+```swift
+viewModel
+    .signIn
+    .sink { result in
+        somethingUnexpected(with: result)
+    }
+    .store(in: &cancellables)
+```
+
+Consider the situation where you happen to be tracking down a bug that occurs when your users attempt to authenticate. 
+Because we have exposed a `Subject` we have to check all the subscriptions to the `signInResult` output _but also_ any subscriptions to the `signIn` input! 
+This makes things harder than they should be...
+
+Let's take a look at a simple approach to get our encapsulation back!:
```

Comment:
I'd avoid the use of the exclamation mark directly before the colon. Not sure if that's allowed punctuation though, so it might just be my own preference.

```suggestion
Let's take a look at a simple approach to get our encapsulation back:
```